File: src/acceptance-test/java/net/slipp/AuthenticationAT.java
package net.slipp;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import net.slipp.qna.FBLogoutPage;
import net.slipp.qna.IndexPage;
import net.slipp.service.user.FixedPasswordGenerator;
import net.slipp.support.AbstractATTest;
import net.slipp.user.ChangePasswordPage;
import net.slipp.user.LoginPage;
import net.slipp.user.ProfilePage;
import net.slipp.web.UserForm;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.openqa.selenium.Alert;

public class AuthenticationAT extends AbstractATTest {
    private IndexPage indexPage;
    
    @Before
    public void setup() {
        super.setup();
        driver.get("http://localhost:8080");
        indexPage = new IndexPage(driver);
    }
    
    @Test
    public void login_after_relogin() throws Exception {
        String firstNickName = environment.getProperty("facebook.nickName1");
        indexPage.loginToFacebook(
                environment.getProperty("facebook.email1"), 
                environment.getProperty("facebook.password1"),
                firstNickName);
        
        FBLogoutPage logoutPage = new FBLogoutPage(driver);
        logoutPage.verifyCurrentLoginUser(firstNickName);
        logoutPage.logout();
        
        String secondNickName = environment.getProperty("facebook.nickName2");
        indexPage.loginToFacebook(
                environment.getProperty("facebook.email2"), 
                environment.getProperty("facebook.password2"), 
                secondNickName);
        logoutPage.goToLogoutPage();
        logoutPage.verifyCurrentLoginUser(secondNickName);
    }
    
    @Test
    public void join() throws Exception {
        join_to_slipp();  
    }
    
    private UserForm join_to_slipp() throws Exception {
        String userId = createUserId();
        String email = System.currentTimeMillis() + environment.getProperty("slipp.email1");
        LoginPage loginPage = indexPage.goLoginPage();
        indexPage = loginPage.join(userId, email);
        assertThat(indexPage.isLoginStatus(), is(true));
        return new UserForm(userId, email);
    }

    private String createUserId() {
        return environment.getProperty("slipp.userId1") + System.currentTimeMillis();
    }
    
    @Test
    public void join_relogin() throws Exception {
        UserForm userForm = join_to_slipp();
        indexPage = indexPage.logout();
        LoginPage loginPage = indexPage.goLoginPage();
        loginPage.loginToSlipp(userForm.getEmail(), FixedPasswordGenerator.DEFAULT_FIXED_PASSWORD);
        assertThat(indexPage.isLoginStatus(), is(true));
    }
    
    @Test
    public void join_change_password_login() throws Exception {
        UserForm userForm = join_to_slipp();
        ProfilePage profilePage = indexPage.goProfilePage();
        profilePage.verifyPageTitle(userForm.getUserId());
        ChangePasswordPage changePasswordPage = profilePage.goChangePasswordPage();
        
        String oldPassword = FixedPasswordGenerator.DEFAULT_FIXED_PASSWORD;
        String newPassword = "newPassword";
        changePasswordPage.changePassword(oldPassword, newPassword);
        Alert alert = driver.switchTo().alert();
        alert.accept();
        LoginPage loginPage = indexPage.goLoginPage();
        loginPage.loginToSlipp(userForm.getEmail(), newPassword);
        assertThat(indexPage.isLoginStatus(), is(true));
    }
    
    @After
    public void teardown() {
        FBLogoutPage logoutPage = new FBLogoutPage(driver);
        logoutPage.logout();
    }
}

File: src/acceptance-test/scala/net/slipp/AuthenticationAT.scala
package net.slipp

import net.slipp.qna.{FBLogoutPage, IndexPage}
import net.slipp.service.user.PasswordGenerator
import net.slipp.support.AbstractATTest
import net.slipp.web.UserForm
import org.hamcrest.CoreMatchers.is
import org.junit.Assert.assertThat
import org.junit.{After, Before, Test}

class AuthenticationAT extends AbstractATTest {
  private var index: IndexPage = null

  @Before override def setup() {
    super.setup()
    driver.get("http://localhost:8080")
    index = new IndexPage(driver)
  }

  @Test def login_after_relogin() {
    val firstNickName = environment.getProperty("facebook.nickName1")
    index.loginToFacebook(environment.getProperty("facebook.email1"), environment.getProperty("facebook.password1"), firstNickName)
    val logoutPage = new FBLogoutPage(driver)
    logoutPage.verifyCurrentLoginUser(firstNickName)
    logoutPage.logout
    val secondNickName = environment.getProperty("facebook.nickName2")
    index.loginToFacebook(environment.getProperty("facebook.email2"), environment.getProperty("facebook.password2"), secondNickName)
    logoutPage.goToLogoutPage
    logoutPage.verifyCurrentLoginUser(secondNickName)
  }

  @Test def login_admin() {
    val indexPage = index.loginToFacebook(environment.getProperty("facebook.email1"), environment.getProperty("facebook.password1"), environment.getProperty("facebook.nickName1"))
    indexPage.verifyAdminPage()
    val logoutPage = new FBLogoutPage(driver)
    logoutPage.logout
    index.loginToFacebook(environment.getProperty("facebook.email2"), environment.getProperty("facebook.password2"), environment.getProperty("facebook.nickName2"))
    indexPage.verifyNonAdminPage()
  }

  @Test def join() {
    join_to_slipp
  }

  private def join_to_slipp = {
    val userId = createUserId
    val email = System.currentTimeMillis + environment.getProperty("slipp.email1")
    val loginPage = index.goLoginPage
    index = loginPage.join(userId, email)
    assertThat(index.isLoginStatus, is(true))
    new UserForm(userId, email)
  }

  private def createUserId = {
    environment.getProperty("slipp.userId1") + System.currentTimeMillis
  }

  @Test def join_relogin() {
    val userForm = join_to_slipp
    index = index.logout
    val loginPage = index.goLoginPage
    loginPage.loginToSlipp(userForm.getEmail, PasswordGenerator.DEFAULT_FIXED_PASSWORD)
    assertThat(index.isLoginStatus, is(true))
  }

  @Test def join_change_password_login() {
    val userForm = join_to_slipp
    val profilePage = index.goProfilePage
    profilePage.verifyPageTitle(userForm.getUserId)

    val changePasswordPage = profilePage.goChangePasswordPage
    val oldPassword = PasswordGenerator.DEFAULT_FIXED_PASSWORD
    val newPassword = "newPassword"
    changePasswordPage.changePassword(oldPassword, newPassword)

    val alert = driver.switchTo.alert
    alert.accept()

    val loginPage = index.goLoginPage
    loginPage.loginToSlipp(userForm.getEmail, newPassword)
    assertThat(index.isLoginStatus, is(true))
  }

  @After def teardown() {
    new FBLogoutPage(driver).logout
  }
}
